packages/state-transition/src/cache/epochCache.ts

```diff
@@ -736,9 +772,52 @@ export class EpochCache {
     return isAggregatorFromCommitteeLength(committee.length, slotSignature);
   }
 
-  addPubkey(index: ValidatorIndex, pubkey: Uint8Array): void {
-    this.pubkey2index.set(pubkey, index);
-    this.index2pubkey[index] = bls.PublicKey.fromBytes(pubkey, CoordType.jacobian); // Optimize for aggregation
+  getPubkey(index: ValidatorIndex): PublicKey {
+    return this.finalizedIndex2pubkey[index];
+  }
+
+  getValidatorIndex(pubkey: Uint8Array | PubkeyHex): ValidatorIndex | undefined {
+    if (this.isAfterEIP6110()) {
+      return (
+        this.finalizedPubkey2index.get(pubkey) ?? this.unfinalizedPubkey2index.get(toMemoryEfficientHexStr(pubkey))
+      );
+    } else {
+      return this.finalizedPubkey2index.get(pubkey);
+    }
+  }
+
+  /**
+   *
+   * Add unfinalized pubkeys
+   *
+   */
+  addPubkey(pubkey: Uint8Array, index: ValidatorIndex): void {
+    if (this.isAfterEIP6110()) {
+      this.unfinalizedPubkey2index = this.unfinalizedPubkey2index.set(toMemoryEfficientHexStr(pubkey), index);
+    } else {
+      this.finalizedPubkey2index.set(pubkey, index);
+      this.finalizedIndex2pubkey[index] = bls.PublicKey.fromBytes(pubkey, CoordType.jacobian);
+    }
+  }
+
+  /**
+   *
+   * Given a validator whose activation_epoch has just been set, we move its pubkey from unfinalized cache to finalized cache
+   *
+   */
+  addFinalizedPubkey(pubkey: Uint8Array, index: ValidatorIndex): void {
+    if (!this.isAfterEIP6110()) {
+      throw new Error("addFInalizedPubkey is not available pre EIP-6110");
+    }
+
+    if (this.finalizedPubkey2index.get(pubkey) !== undefined) {
+      return; // Repeated insert. Should not happen except 6110 activation
```

Comment:
This guard could be more thorough, checking that:
```ts
const existingIndex = this.finalizedPubkey2index.get(pubkey);
if (existingIndex != undefined) {
  if (existingIndex == index) {
    // duplicate insertion, TODO: explain in detail how this can happen
  } else {
    // attempt to insert the same pubkey with different index, should never happen, what to do?
    // throwing an error with brick Lodestar nodes if it ever happens, but otherwise it can allow a
    // very bad bug to happen
    throw Error("inserted existing pubkey into finalizedPubkey2index cache with a different index")
  }
}
```